File: src/go/server/fcmeans_server.go
package main

import (
	"bytes"
	"log"
	"math"
	"os"
	"time"

	"encoding/gob"
	"encoding/json"
	"math/rand"

	"fcmeans/common"

	"github.com/ergo-services/ergo/etf"
	"github.com/shirou/gopsutil/cpu"
	"github.com/shirou/gopsutil/mem"
	"gonum.org/v1/gonum/mat"
)

type FCMeansServer struct {
	target_feature    int
	max_number_rounds int
	num_clusters      int
	epsilon           int
	num_features      int
	cluster_centers   [][][]float64
	*FLExperiment
	currentRound int
}

type FLExperiment struct {
	_global_model_parameters    [][]float64
	_client_configuration       map[string]interface{}
	_max_number_rounds          int
	_client_ids                 interface{}
	_type_of_termination        string
	_termination_threshold      float64
	_monitored_metric           interface{}
	_type_of_client_selection   interface{}
	_client_ratio               interface{}
	_client_values              interface{}
	_step_wise_client_selection bool
	_client_selection_threshold interface{}
	_round_selected_ids         []int
	_latency                    int
	_calls_list                 []etf.Tuple
	_validate                   bool
	experiment_history          interface{}
	_latency_required           bool
}

func (s *FCMeansServer) Init_server(experiment, json_str_config, bb string, fp common.FedLangProcess) etf.Term {
	gob.Register([][]float64{})
	byteSlice := []byte(bb)
	client_ids := make([]int, len(byteSlice))
	for i, b := range byteSlice {
		client_ids[i] = int(b)
	}
	var experiment_config map[string]interface{}

	// Parse the JSON string
	err := json.Unmarshal([]byte(json_str_config), &experiment_config)
	if err != nil {
		log.Fatal(err)
	}
	log.Printf("experiment = %#v, client_ids = %#v, experiment_config = %#v\n", experiment, client_ids, experiment_config)
	parameters := experiment_config["parameters"].(map[string]interface{})
	log.Printf("paramters->numFeatures = %#v\n", parameters["numFeatures"])
	n_features := int(parameters["numFeatures"].(float64))
	seed := int(parameters["seed"].(float64))
	n_clusters := int(parameters["numClusters"].(float64))
	s.target_feature = int(parameters["targetFeature"].(float64))
	s.max_number_rounds = int(experiment_config["maxNumberOfRounds"].(float64)) // WARN: max_number_rounds should be in parameters??
	lambda_factor := int(parameters["lambdaFactor"].(float64))
	stop_condition_threshold := experiment_config["stopConditionThreshold"]
	log.Printf("n_features = %#v, seed = %#v, n_clusters = %#v, target_feature = %#v, max_number_rounds = %#v, lambda_factor = %#v, stop_condition_threshold = %#v\n", n_features, seed, n_clusters, s.target_feature, s.max_number_rounds, lambda_factor, stop_condition_threshold)
	if stop_condition_threshold != nil {
		s.epsilon = int(stop_condition_threshold.(float64))
	} else {
		s.epsilon = -1 // HACK: check logic behind stop_condition using epsilon
		log.Println("stop_condition_threshold is nil")
	}
	log.Printf("n_clusters = %#v, epsilon = %#v, max_number_rounds = %#v, n_features = %#v", n_clusters, s.epsilon, s.max_number_rounds, n_features)
	min_num_clients := int(experiment_config["minNumberClients"].(float64))
	log.Printf("min_num_clients = %#v\n", min_num_clients)
	flexperiment := FLExperiment{
		_max_number_rounds: s.max_number_rounds,
		_client_ids:        client_ids,
		_client_configuration: map[string]interface{}{
			"lambdaFactor":  lambda_factor,
			"numClients":    min_num_clients,
			"targetFeature": s.target_feature,
			"do_sleep":      true,
		},
		_latency_required: true,
		_calls_list:       nil,
		_validate:         false,
	}
	log.Printf("flexperiment = %#v\n", flexperiment)
	rand.Seed(int64(seed))

	centers := make([][]float64, n_clusters)
	for i := 0; i < n_clusters; i++ {
		centers[i] = make([]float64, n_features)
		for j := 0; j < n_features; j++ {
			centers[i][j] = rand.Float64()
		}
	}
	// log.Printf("centers = %#v\n", centers)
	s.num_clusters = n_clusters
	s.num_features = n_features
	cluster_centers := make([][][]float64, 0)
	cluster_centers = append(cluster_centers, centers)
	s.cluster_centers = cluster_centers
	_ = make([]float64, 0)

	centerList := make([][]float64, len(centers))
	for i, center := range centers {
		centerList[i] = center
	}
	log.Printf("centerList = %#v\n", centerList)

	flexperiment._global_model_parameters = centerList

	flexperiment._type_of_termination = "custom"
	flexperiment._type_of_client_selection = "probability"
	flexperiment._client_ratio = 1
	flexperiment._client_values = nil
	flexperiment._step_wise_client_selection = false
	flexperiment._client_selection_threshold = nil
	s.FLExperiment = &flexperiment
	s.currentRound = 0

	clientConfig, _, _, _, _, callsList := flexperiment.get_initialization()
	log.Printf("clientConfig = %#v\n", clientConfig)

	clientConfigurationStr, err := json.Marshal(clientConfig)
	if err != nil {
		log.Fatalf("Error marshaling client config: %v", err)
	}
	atom := etf.Atom("fl_server_ready")
	log.Printf("atom = %#v\n", atom)
	pid := fp.Own_pid
	log.Printf("pid = %#v\n", pid)
	// log.Printf("clientConfigurationStr = %#v\n", clientConfigurationStr)
	log.Printf("callsList = %#v\n", callsList)
	msg := etf.Tuple{atom, pid, clientConfigurationStr, callsList}
	log.Printf("sending message = %#v\n", msg)
	// err = s.Process.Send(
	// 	gen.ProcessID{Name: s.Erl_worker_mailbox, Node: s.Erl_client_name},
	// 	msg,
	// )
	// if err != nil {
	// 	log.Fatalf("Error sending message: %v", err)
	// 	panic(err)
	// }
	log.Printf("message sent = %#v\n", msg)
	return msg
}

func (e *FLExperiment) get_initialization() (map[string]interface{}, int, []byte, float64, int, []etf.Tuple) {
	typeOfTermination := []byte(e._type_of_termination)

	// Handling the callsList logic
	if len(e._calls_list) == 0 {
		if e._validate {
			e._calls_list = append(e._calls_list, etf.Tuple{[]byte("standard"), []byte("two_step")})
		} else {
			e._calls_list = append(e._calls_list, etf.Tuple{[]byte("standard"), []byte("one_step")})
		}
	}

	// Returning multiple values
	return e._client_configuration, e._max_number_rounds, typeOfTermination, e._termination_threshold, e._latency, e._calls_list
}

var startFlTime = time.Time{}

func encodeToBytes(data interface{}) ([]byte, error) {
	var buf bytes.Buffer
	encoder := gob.NewEncoder(&buf)
	err := encoder.Encode(data)
	if err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}

func decodeFromBytes(data []byte, v interface{}) error {
	buf := bytes.NewBuffer(data)
	decoder := gob.NewDecoder(buf)
	return decoder.Decode(v)
}

func (s *FCMeansServer) Start_round(round_mail_box, experiment string, round_number int, fp common.FedLangProcess) etf.Term {
	// log.Printf("round_mail_box = %#v, experiment = %#v, round_number = %#v\n", round_mail_box, experiment, round_number)
	if startFlTime == (time.Time{}) {
		startFlTime = time.Now()
	}
	log.Printf("start_fl_time = %#v\n", startFlTime)
	result := s.FLExperiment._global_model_parameters
	// log.Printf("start round result = %#v\n", result)
	client_ids := s.FLExperiment._client_ids
	// log.Printf("before sending result to (%#v, %#v)", s.Erl_worker_mailbox, s.Erl_client_name)

	// The serialized data is now in buffer.Bytes()
	tt, err := encodeToBytes(result)
	if err != nil {
		panic(err)
	}

	return etf.Tuple{etf.Atom("start_round_ok"), tt, client_ids}

	// log.Printf("after sending (%#v, %#v) ! (start_round_ok)", s.erl_worker_mailbox, s.Erl_client_name)
}

func (fl *FLExperiment) get_step_data() (interface{}, []int) {
	var step_selected_ids []int
	if !fl._step_wise_client_selection {
		step_selected_ids = fl._round_selected_ids
	} else {
		// TODO: implement perform_client_selection
		//step_selected_ids = fl._perform_client_selection()
		panic("Not implemented")
	}
	fl._round_selected_ids = step_selected_ids
	return fl._global_model_parameters, fl._round_selected_ids
}
func flatten(input [][]float64) []float64 {
	var result []float64
	for _, innerSlice := range input {
		result = append(result, innerSlice...)
	}
	return result
}

func (s *FCMeansServer) Process_server(round_mail_box string, experiment string, config_file int, client_responses etf.List, fp common.FedLangProcess) etf.Term {
	log.Printf("Starting process_server ...")
	// log.Printf("round_mail_box = %#v, experiment = %#v, config_file = %#v, client_responses = %#v\n", round_mail_box, experiment, config_file, client_responses)

	rand.Seed(time.Now().UnixNano())
	const low = 2
	const high = 4
	// time_to_sleep := low + rand.Float64()*(high-low)
	// time.Sleep(time.Duration(time_to_sleep) * time.Second)

	// log.Printf("start process_server, experiment = %s, round_mail_box = %s, len(client_responses) = %d\n", experiment, round_mail_box, len(client_responses))

	type clientDataType struct {
		clientId int
		result   struct {
			U  []float64
			Ws [][]float64
		}
	}
	var data []clientDataType
	// etf.List{etf.Tuple{0, []uint8{
	for _, clientResponse := range client_responses {
		clientResponseSlice, ok := clientResponse.(etf.Tuple)
		if !ok {
			panic("Error: clientResponse is not of type etf.Tuple")
		}
		if len(clientResponseSlice) < 2 {
			panic("Error: clientResponseSlice is < 2")
		}
		var decodedResult struct {
			U  []float64
			Ws [][]float64
		}
		// ([1.0, 2.0, ...], [[1.0, 2.0, ...], [1.0, 2.0, ...], ...])
		// decoder := pickle.NewDecoder(bytes.NewReader(clientResponseSlice[1].([]byte)))
		// decodedResult_tmp, err := decoder.Decode()
		err := decodeFromBytes(clientResponseSlice[1].([]byte), &decodedResult)
		if err != nil {
			panic(err)
		}

		responseData := clientDataType{clientResponseSlice[0].(int), decodedResult}
		data = append(data, responseData)
	}

	for i, d := range data {
		log.Printf("i = %d, data[0] = %v", i, d.clientId)
	}

	var cl_resp []interface{}
	for _, cr := range data {
		cl_resp = append(cl_resp, cr.result)
	}
	// log.Printf("cl_resp = %#v\n", cl_resp)

	s.currentRound += 1
	num_clients := len(client_responses)
	log.Printf("num_clients = %#v\n", num_clients)

	uList := make([]float64, s.num_clusters)
	wsList := make([][]float64, s.num_clusters)
	for i := range wsList {
		wsList[i] = make([]float64, s.num_features)
	}

	for client_idx := 0; client_idx < num_clients; client_idx++ {
		response := data[client_idx].result
		us := response.U
		wss := response.Ws
		for i := 0; i < s.num_clusters; i++ {
			var clientWs *mat.Dense
			clientU := us[i]
			log.Printf("clientU = %#v\n", clientU)
			ws := wss[i]
			log.Printf("ws = %#v\n", ws)
			clientWs = mat.NewDense(1, len(ws), nil)
			for j, w := range ws {
				clientWs.Set(0, j, w)
			}
			log.Printf("clientWs = %#v\n", clientWs)
			uList[i] += clientU
			for j := 0; j < clientWs.RawMatrix().Cols; j++ {
				wsList[i][j] += clientWs.At(0, j)
			}
		}
	}
	var newClusterCenters [][]float64
	prevClusterCenters := s.cluster_centers[len(s.cluster_centers)-1]

	for i := 0; i < s.num_clusters; i++ {
		u := uList[i]
		ws := wsList[i]
		var center []float64
		if u == 0 {
			center = prevClusterCenters[i]
		} else {
			center = make([]float64, len(ws))
			for j := range ws {
				center[j] = ws[j] / float64(u)
			}
		}
		newClusterCenters = append(newClusterCenters, center)
	}
	s.FLExperiment._global_model_parameters = newClusterCenters
	s.cluster_centers = append(s.cluster_centers, newClusterCenters)

	step_data, _ := s.FLExperiment.get_step_data() // TODO add client ids
	centersR := mat.NewDense(len(newClusterCenters), len(newClusterCenters[0]), flatten(newClusterCenters))
	centersR1 := mat.NewDense(len(prevClusterCenters), len(prevClusterCenters[0]), flatten(prevClusterCenters))
	diffMatrix := mat.NewDense(centersR.RawMatrix().Rows, centersR.RawMatrix().Cols, nil)
	diffMatrix.Sub(centersR, centersR1)

	v, _ := mem.VirtualMemory()
	total_memory := v.Total
	used_memory := v.Used
	memory_usage_percentage := math.Round((float64(used_memory)/float64(total_memory))*100) / 100

	cpu_percentages, _ := cpu.Percent(time.Second, false)

	metricsMessage := map[string]interface{}{
		"timestamp": time.Now().Unix(),
		"round":     s.currentRound,
		"hostMetrics": map[string]float64{
			"cpuUsagePercentage":    cpu_percentages[0],
			"memoryUsagePercentage": memory_usage_percentage,
		},
		"modelMetrics": map[string](mat.Matrix){
			"FRO": diffMatrix,
		},
	}
	metricsMessageBytes, _ := json.Marshal(metricsMessage)
	// s.Process.Send(
	// 	gen.ProcessID{Name: s.Erl_worker_mailbox, Node: s.Erl_client_name},
	// 	etf.Tuple{etf.Atom("process_server_ok"), buffer.Bytes(), metricsMessageBytes},
	// )
	sd_bytes, err := encodeToBytes(step_data)
	if err != nil {
		panic(err)
	}
	return etf.Tuple{etf.Atom("process_server_ok"), sd_bytes, metricsMessageBytes}
}

func (s *FCMeansServer) Finish(fp common.FedLangProcess) {
	log.Printf("DESTROY")
	os.Exit(0)
}

func main() {
	go_node_id := os.Args[1]         // go_c0ecdfb7-00f1-4270-8e46-d835bd00f153@127.0.0.1
	erl_client_name := os.Args[2]    // director@127.0.0.1
	erl_worker_mailbox := os.Args[3] // mboxserver_c0ecdfb7-00f1-4270-8e46-d835bd00f153
	erl_cookie := os.Args[4]         // cookie_123456789
	experiment_id := os.Args[5]      // c0ecdfb7-00f1-4270-8e46-d835bd00f153

	log.Printf("gorlang_node_id = %v, erl_client_name = %v, erl_worker_mailbox = %v, erl_cookie = %v, experiment_id = %v\n", go_node_id, erl_client_name, erl_worker_mailbox, erl_cookie, experiment_id)

	logFileName := os.Getenv("FL_CLIENT_LOG_FOLDER") + "/" + os.Getenv("FL_CLIENT_ID") + ".log"
	if logFileName == "" {
		logFileName = "default.log"
	}
	logFile, err := os.OpenFile(logFileName, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0666)
	if err != nil {
		log.Fatalf("Failed to open log file: %v", err)
	}
	defer logFile.Close()
	log.SetOutput(logFile)

	common.StartProcess[FCMeansServer](go_node_id, erl_cookie, erl_client_name, erl_worker_mailbox, experiment_id)
}

File: src/python_server/fcmeans_server.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# Copyright (C) 2024 AI&RD Research Group, Department of Information Engineering, University of Pisa
# SPDX-License-Identifier: Apache-2.0

import os
import sys
import traceback
import time
import numpy as np
import _pickle as cPickle
from term import Atom
from typing import Final
import python_common.common_utils as fl_utils
from python_common.common_utils import logger_info, logger_error
from timeit import default_timer as timer
import json
import psutil

from python_common.fedlang_process import FedLangProcess, start_process

LOG_FOLDER_PATH: Final[str] = os.environ.get("FL_CLIENT_LOG_FOLDER")
FROBENIUS_NORM: Final[str] = 'fro'

start_fl_time = None

process = psutil.Process()


def norm_fro(u: np.ndarray):
    return np.linalg.norm(u, ord=FROBENIUS_NORM)


class FCMeansServerProcess(FedLangProcess):

    def __init__(self, **kwargs):
        super().__init__(**kwargs)

    def init_server(self, experiment: str, json_str_config: str, bb):
        try:
            bb = bytes(bb, encoding='utf8')
            client_ids = list(bb)
            experiment_config = json.loads(json_str_config)
            logger_info(f"experiment = {experiment}, client_ids = {client_ids}, experiment_config = {experiment_config}")
            parameters = experiment_config.get("parameters")
            n_features = int(parameters.get("numFeatures"))
            seed = int(parameters.get("seed"))
            n_clusters = int(parameters.get("numFeatures"))
            self.target_feature = parameters.get("targetFeature")
            self.max_number_rounds = int(experiment_config.get("maxNumberOfRounds"))
            lambda_factor = int(parameters.get("lambdaFactor"))
            _stop_condition_threshold = experiment_config.get("stopConditionThreshold")
            if _stop_condition_threshold is not None:
                self.epsilon = int(_stop_condition_threshold)
            else:
                self.epsilon = None
            logger_info(f"n_clusters = {n_clusters}, epsilon = {self.epsilon}, max_number_rounds = {self.max_number_rounds}, n_features = {n_features}")
            min_num_clients = int(experiment_config.get("minNumberClients"))
            experiment = fl_utils.FLExperiment(self.max_number_rounds, client_ids)
            experiment.set_client_configuration({
                'lambdaFactor': lambda_factor,
                'numClients': min_num_clients,
                'targetFeature': self.target_feature,
                'do_sleep': True
            })
            experiment.latency_required(True)
            np.random.seed(seed)
            centers = np.random.rand(n_clusters, n_features)
            self.num_clusters = n_clusters
            self.num_features = n_features
            cluster_centers = list()
            cluster_centers.append(centers)
            self.cluster_centers = cluster_centers
            self.fnorms = list()
            experiment.set_global_model_parameters([center.tolist() for center in centers])
            logger_info(f'centers = {centers}')
            experiment.set_type_of_termination('custom')
            experiment.set_type_of_client_selection('probability', 1.0)
            self.experiment = experiment
            self.current_round = 0
            client_config, max_num_rounds, termination_type, threshold_value, latency, calls_list = self.experiment.get_initialization()
            client_configuration_str = json.dumps(client_config)
            logger_info(f"before sending  ({erl_worker_mailbox}, {erl_client_name} )! (fl_server_ready)")
            self.get_node().send_nowait(sender=self.pid_,
                                        receiver=(Atom(erl_client_name), Atom(erl_worker_mailbox)),
                                        message=(Atom('fl_server_ready'), self.pid_, client_configuration_str, calls_list))
            logger_info(f"after sending  ({erl_worker_mailbox}, {erl_client_name} )! (fl_server_ready)")
        except:
            logger_error(traceback.format_exc())

    def start_round(self, round_mail_box, experiment, round_number):
        global start_fl_time
        if start_fl_time is None:
            start_fl_time = timer()
        result, client_ids = self.experiment.get_round_parameters()
        logger_info(f"start_round result = {result}")
        logger_info(f"before sending  ({erl_worker_mailbox}, {erl_client_name} )! (start_round_ok)")
        tt = cPickle.dumps(result)
        self.round_number = round_number
        self.get_node().send_nowait(sender=self.pid_,
                                    receiver=(Atom(erl_client_name), Atom(round_mail_box)),
                                    message=(Atom('start_round_ok'), tt, client_ids))
        logger_info(f"after sending  ({erl_worker_mailbox}, {erl_client_name} )! (start_round_ok)")

    #@profile(stream=open(str(os.environ.get("METRIC_FILE")), 'w'))
    def process_server(self, round_mail_box, experiment, config_file, client_responses):
        start_time = timer()
        try:
            time_to_sleep = np.random.uniform(2, 4)
            #time.sleep(time_to_sleep)
            logger_info(f"start process_server, experiment = {experiment}, round_mail_box = {round_mail_box}, len(client_responses) = {len(client_responses)}")
            data = [(client_response[0], cPickle.loads(bytes(client_response[1]))) for client_response in
                    client_responses]
            for i in range(len(data)):
                fl_utils.logger.info(f'i = {i}, data[0] = {data[i][0]}')

            client_responses = [cr[1] for cr in data]

            self.current_round = self.current_round + 1
            num_clients = len(client_responses)

            u_list = [0] * self.num_clusters
            ws_list = [[0] * self.num_features for i in range(self.num_clusters)]

            for client_idx in range(num_clients):
                # remember the response is a list of tuples where each tuple represents the (u, ws) for each cluster
                response = client_responses[client_idx]
                for i in range(self.num_clusters):
                    client_u = response[0][i]
                    client_ws = response[1][i] if response[1][i] is np.array else np.array(response[1][i])
                    u_list[i] = u_list[i] + client_u
                    ws_list[i] = ws_list[i] + client_ws

            new_cluster_centers = []
            prev_cluster_centers = self.cluster_centers[-1]

            for i in range(self.num_clusters):
                u = u_list[i]
                ws = ws_list[i]
                if (u == 0):
                    center = prev_cluster_centers[i]
                else:
                    center = ws / (u * 1.0)
                new_cluster_centers.append(np.array(center))

            self.experiment.set_global_model_parameters([centers.tolist() for centers in new_cluster_centers])
            self.cluster_centers.append(new_cluster_centers)

            data, client_ids = self.experiment.get_step_data([centers.tolist() for centers in new_cluster_centers])
            #logger_info(f"process_server, data = {data}, client_ids = {client_ids}")

            centers_r = np.array(self.cluster_centers[-1])
            centers_r_1 = np.array(self.cluster_centers[-2])
            tot_diff_sum = norm_fro(centers_r - centers_r_1)

            total_memory, used_memory, free_memory = map(int, os.popen('free -t -m').readlines()[-1].split()[1:])
            memory_usage_percentage = round((used_memory/total_memory) * 100, 2)
            cpu_usage_percentage = psutil.cpu_percent()
            metrics_message = {
                "timestamp": int(time.time()), "type": "strategy_server_metrics",
                "round": self.current_round,
                "hostMetrics": {
                                "cpuUsagePercentage": cpu_usage_percentage,
                                "memoryUsagePercentage": memory_usage_percentage
                },
                "modelMetrics": {"FRO": tot_diff_sum}
            }

            self.get_node().send_nowait(sender=self.pid_,
                                        receiver=(Atom(erl_client_name), Atom(round_mail_box)),
                                        message=(Atom('process_server_ok'), cPickle.dumps(data), json.dumps(metrics_message)))
            logger_info(f"end process_server, experiment = {experiment}, round = {config_file}")

        except:
            logger_error(traceback.format_exc())

    def next_round(self):
        try:
            num_centers = len(self.cluster_centers)
            if num_centers > 1:
                centers_r = np.array(self.cluster_centers[-1])
                centers_r_1 = np.array(self.cluster_centers[-2])
                tot_diff_sum = norm_fro(centers_r - centers_r_1)
                self.fnorms.append(tot_diff_sum)

            to_return = self.round_number < self.max_number_rounds
            self.get_node().send_nowait(sender=self.pid_,
                                        receiver=(Atom(erl_client_name), Atom(erl_worker_mailbox)),
                                        message=(Atom('next_round_ok'), to_return))
            logger_info(f"end next_round, round_number = {self.round_number}, to_return = {to_return}")
        except:
            logger_error(traceback.format_exc())

    def get_current_metric_value(self):
        try:
            num_centers = len(self.cluster_centers)
            tot_diff_sum = self.epsilon
            if num_centers > 1:
                centers_r = np.array(self.cluster_centers[-1])
                centers_r_1 = np.array(self.cluster_centers[-2])
                tot_diff_sum = norm_fro(centers_r - centers_r_1)

            self.get_node().send_nowait(sender=self.pid_,
                                        receiver=(Atom(erl_client_name), Atom(erl_worker_mailbox)),
                                        message=(Atom('get_current_metric_value_ok'), float(tot_diff_sum)))
            logger_info(f"end get_current_metric_value, round_number = {self.round_number}, "
                        f"tot_diff_sum = {tot_diff_sum}, type(tot_diff_sum) = {type(tot_diff_sum)}")
        except:
            logger_error(traceback.format_exc())

    def finish(self):
        logger_info('DESTROY')
        self.exit()
        sys.exit(0)


if __name__ == "__main__":
    pyrlang_node_id = sys.argv[1]
    erl_client_name = sys.argv[2]
    erl_worker_mailbox = sys.argv[3]
    erl_cookie = sys.argv[4]
    experiment_id = sys.argv[5]
    runtime_file: str = os.path.join(LOG_FOLDER_PATH, f"time_by_method_aggregator_{experiment_id}.log")
    logger_info(f'pyrlang_node_id = {pyrlang_node_id}, '
                f'erl_client_name = {erl_client_name}, '
                f'erl_worker_mailbox = {erl_worker_mailbox}, '
                f'erl_cookie = {erl_cookie}')
    start_process(FCMeansServerProcess, pyrlang_node_id, erl_cookie, erl_client_name, erl_worker_mailbox)
